File: front/sync_theme_full.js
const fs = require('fs');
const path = require('path');

const ROOT = __dirname;
const XML_PATH = path.join(ROOT, "Default-theme.xml");
const MANIFEST_PATH = path.join(ROOT, "theme_templates.json");
const RPG_CSS_PATH = path.join(ROOT, "..", "back", "forum", "rpg_custom.css");
const MINIMAL_CSS_PATH = path.join(ROOT, "templates", "mybb", "global", "mybb-minimal.css");

if (!fs.existsSync(XML_PATH)) {
    console.error("Default-theme.xml not found.");
    process.exit(1);
}

let xml = fs.readFileSync(XML_PATH, 'utf8');
const manifest = JSON.parse(fs.readFileSync(MANIFEST_PATH, 'utf8'));

for (const [name, rel] of Object.entries(manifest)) {
    const filePath = path.join(ROOT, rel.replace(/\//g, path.sep));
    if (!fs.existsSync(filePath)) {
        console.log(`SKIP ${name} (missing ${rel})`);
        continue;
    }
    let html = fs.readFileSync(filePath, 'utf8').replace(/\r\n/g, "\n");
    // Escape special regex characters in the template name
    const escapedName = name.replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&');
    const pattern = new RegExp(`(<template\\s+name="${escapedName}"[^>]*><!\\[CDATA\\[)(.*?)(\\]\\]><\\/template>)`, 'gs');
    
    if (!pattern.test(xml)) {
        console.log(`WARN ${name} not in XML`);
        continue;
    }
    
    xml = xml.replace(pattern, (match, p1, p2, p3) => {
        return p1 + html + "\n\t\t" + p3;
    });
    console.log(`OK ${name}`);
}

if (fs.existsSync(RPG_CSS_PATH)) {
    const css = fs.readFileSync(RPG_CSS_PATH, 'utf8').replace(/\r\n/g, "\n");
    const marker = "/* RPG Premium Modern Theme */";
    const cssPattern = /(<stylesheet\s+name="global\.css"[^>]*><!\[CDATA\[)(.*?)(<\/stylesheet>)/gs;
    
    let baseCss = "";
    if (fs.existsSync(MINIMAL_CSS_PATH)) {
        baseCss = fs.readFileSync(MINIMAL_CSS_PATH, 'utf8').replace(/\r\n/g, "\n");
        if (baseCss && !baseCss.endsWith("\n")) {
            baseCss += "\n";
        }
    }
    
    xml = xml.replace(cssPattern, (match, p1, p2, p3) => {
        return p1 + baseCss + marker + "\n" + css + "]]>\n\t\t" + p3;
    });
    console.log("OK global.css");
}

fs.writeFileSync(XML_PATH, xml, 'utf8');
console.log("Default-theme.xml updated successfully.");

File: front/sync_theme_full.py
#!/usr/bin/env python3
"""Sync front/templates + rpg_custom.css into Default-theme.xml."""
import json
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parent
XML = ROOT / "Default-theme.xml"
MANIFEST = ROOT / "theme_templates.json"
RPG_CSS = ROOT.parent / "back" / "forum" / "rpg_custom.css"
MINIMAL_CSS = ROOT / "templates" / "mybb" / "global" / "mybb-minimal.css"

xml = XML.read_text(encoding="utf-8")
templates = json.loads(MANIFEST.read_text(encoding="utf-8"))

for name, rel in templates.items():
    path = ROOT / rel.replace("/", "\\")
    if not path.is_file():
        print(f"SKIP {name} (missing {rel})")
        continue
    html = path.read_text(encoding="utf-8").replace("\r\n", "\n")
    pattern = rf'(<template\s+name="{re.escape(name)}"[\s\S]*?>\s*<!\[CDATA\[)(.*?)(\]\]>\s*</template>)'
    if not re.search(pattern, xml, re.DOTALL):
        print(f"WARN {name} not in XML")
        continue
    xml = re.sub(pattern, lambda m: m.group(1) + html + "\n\t\t" + m.group(3), xml, count=1, flags=re.DOTALL)
    print(f"OK {name}")

if RPG_CSS.is_file():
    css = RPG_CSS.read_text(encoding='utf-8-sig').replace('\r\n', '\n').lstrip('\ufeff')
    marker = "/* RPG Premium Modern Theme */"
    css_pattern = r'(<stylesheet\s+name="global\.css"[^>]*><!\[CDATA\[)(.*?)(</stylesheet>)'

    def repl_css(m):
        base = MINIMAL_CSS.read_text(encoding="utf-8").replace("\r\n", "\n") if MINIMAL_CSS.is_file() else ""
        if base and not base.endswith("\n"):
            base += "\n"
        return m.group(1) + base + marker + "\n" + css + "]]>\n\t\t" + m.group(3)

    xml = re.sub(css_pattern, repl_css, xml, count=1, flags=re.DOTALL)
    print("OK global.css")

XML.write_text(xml, encoding="utf-8")
print("Default-theme.xml updated.")
